sync/syncer.go

```diff
@@ -291,149 +411,133 @@ func (s *Syncer) Txs(mb *types.Block) ([]*types.SerializableTransaction, error)
 
 	//map and sort txs
 	if len(missing) > 0 {
-		for out := range s.fetchWithFactory(TxReqFactory(missinDB), 1) {
+		for out := range s.fetchWithFactory(NewNeighborhoodWorker(s, 1, TxReqFactory(missinDB))) {
 			ntxs := out.([]types.SerializableTransaction)
 			for _, tx := range ntxs {
-				txMap[types.GetTransactionId(&tx)] = &tx
+				if valid, err := s.ValidateTx(&tx); !valid || err != nil {
+					id := types.GetTransactionId(&tx)
+					s.Warning("tx %v not valid %v", hex.EncodeToString(id[:]), err)
+					continue
+				}
+				poolTxs[types.GetTransactionId(&tx)] = &tx
 			}
 		}
 	}
 
-	txs := make([]*types.SerializableTransaction, 0, len(mb.TxIds))
-	for _, t := range mb.TxIds {
-		if tx, ok := txMap[t]; ok {
+	txs := make([]*types.SerializableTransaction, 0, len(txids))
+	for _, id := range txids {
+		if tx, ok := poolTxs[id]; ok {
 			txs = append(txs, tx)
-		} else if tx, ok := dbTxs[t]; ok {
+		} else if tx, ok := dbTxs[id]; ok {
 			txs = append(txs, tx)
 		} else {
-			return nil, errors.New(fmt.Sprintf("could not fetch tx %v", hex.EncodeToString(t[:])))
+			return nil, errors.New(fmt.Sprintf("could not fetch tx %v", hex.EncodeToString(id[:])))
 		}
 	}
 	return txs, nil
 }
 
-func (s *Syncer) ATXs(mb *types.Block) (atxs []*types.ActivationTx, associated *types.ActivationTx, err error) {
+func (s *Syncer) syncAtxs(atxIds []types.AtxId) (atxs []*types.ActivationTx, err error) {
 
 	//look in pool
-	atxMap := make(map[types.AtxId]*types.ActivationTx, len(atxs))
+	poolAtxs := make(map[types.AtxId]*types.ActivationTx, len(atxs))
 	missingInPool := make([]types.AtxId, 0, len(atxs))
-	for _, t := range mb.AtxIds {
+	for _, t := range atxIds {
 		id := t
 		if x := s.atxpool.Get(id); x != nil {
 			atx := x.(*types.ActivationTx)
 			if atx.Nipst == nil {
-				s.Warning("atx %v nipst not found ", hex.EncodeToString(id.Bytes()))
+				s.Warning("atx %v nipst not found ", id.ShortId())
 				missingInPool = append(missingInPool, id)
 				continue
 			}
-			s.Info("found atx, %v in atx pool", hex.EncodeToString(id.Bytes()))
-			atxMap[id] = atx
+			s.Info("found atx, %v in atx pool", id.ShortId())
+			poolAtxs[id] = atx
 		} else {
-			s.Warning("atx %v not in atx pool", hex.EncodeToString(id.Bytes()))
+			s.Warning("atx %v not in atx pool", id.ShortId())
 			missingInPool = append(missingInPool, id)
 		}
 	}
 
 	//look in db
-	dbAtxs, missing := s.GetATXs(missingInPool)
-
-	fetchAssociated := false
-	if mb.ATXID != *types.EmptyAtxId {
-		if _, err := s.GetAtx(mb.ATXID); err != nil {
-			fetchAssociated = true
-			missing = append(missing, mb.ATXID)
-		}
-	}
+	dbAtxs, missingInDb := s.GetATXs(missingInPool)
 
 	//map and sort atxs
-	if len(missing) > 0 {
-		output := s.fetchWithFactory(ATxReqFactory(missing), 1)
+	if len(missingInDb) > 0 {
+
+		output := s.fetchWithFactory(NewNeighborhoodWorker(s, 1, ATxReqFactory(missingInDb)))
```

Comment:
What are you doing with the ATXs that you just fetched? Shouldn't you store them in the DB/pool?

Reply:
yes, but it happens in the caller to this method